contracts/verifying/VerifyingSingletonPaymasterV2.sol

```diff
@@ -1,9 +1,386 @@
-// SPDX-License-Identifier: MIT
-pragma solidity ^0.8.17;
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity 0.8.17;
 
-// todo Add pricing
-// Instead of paymaster helpers write parsePaymasterData methods
+/* solhint-disable reason-string */
+/* solhint-disable no-inline-assembly */
+import {ECDSA} from "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import {ReentrancyGuard} from "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+import {UserOperation, UserOperationLib} from "@account-abstraction/contracts/interfaces/UserOperation.sol";
+import "../BasePaymaster.sol";
+import {VerifyingPaymasterErrors} from "../common/Errors.sol";
 
-contract VerifyingSingletonPaymasterV2 {
+contract VerifyingSingletonPaymasterV2 is
+    BasePaymaster,
+    ReentrancyGuard,
+    VerifyingPaymasterErrors
+{
+    using ECDSA for bytes32;
+    using UserOperationLib for UserOperation;
 
+    uint32 private constant PRICE_DENOMINATOR = 1e6;
+
+    // Review if we only wanna encode until signature and have signature offset
+    // paymasterAndData: concat of [paymasterAddress(address), abi.encode(paymasterId, validUntil, validAfter, priceMarkup): makes up 32*4 bytes, signature]
+    uint256 private constant VALID_PND_OFFSET = 20;
+
+    uint256 private constant SIGNATURE_OFFSET = 148;
+
+    // Gas used in EntryPoint._handlePostOp() method (including this#postOp() call)
+    uint256 private unaccountedEPGasOverhead;
+    mapping(address => uint256) public paymasterIdBalances;
+
+    address public verifyingSigner;
+
+    uint32 public fixedPriceMarkup; // immutable? constant? 1.1e6
+
+    event EPGasOverheadChanged(
+        uint256 indexed _oldValue,
```

Comment:
Would be great if we follow the structure where events, errors and structs are moved to the interface

Reply:
ok would move this and update.

Reply:
done